src/misc/paper-write.ts

```diff
@@ -0,0 +1,321 @@
+    const inventory = event.player.inventory as PlayerInventory;
+    const offHand = inventory.itemInOffHand;
+    if (offHand.type == Material.FEATHER) return;
+    if (playersWriting.has(event.player)) return;
+    // Check if the paper already has something written on it
+    if (event.item.itemMeta.lore) {
+      event.player.sendMessage(text('Paperilla lukee:'));
+      event.player.sendMessage(
+        color(ChatColor.WHITE, text(event.item.itemMeta.lore[0])),
+      );
+    }
+  },
+);
+
+PaperSealed.event(
+  PlayerInteractEvent,
+  (event) => event.item,
+  async (event) => {
+    if (!event.item) return;
+    const inventory = event.player.inventory as PlayerInventory;
+    const offHand = inventory.itemInOffHand;
+    if (offHand.type == Material.FEATHER) return;
+    if (playersWriting.has(event.player)) return;
+    // Check if the paper already has something written on it
+    if (event.item.itemMeta.lore) {
+      event.player.sendMessage(text('Paperilla lukee:'));
+      event.player.sendMessage(
+        color(ChatColor.WHITE, text(event.item.itemMeta.lore[0])),
+      );
+      event.player.sendMessage(
+        color(
+          ChatColor.RED,
+          text(event.item.itemMeta.lore[1] + '' + event.item.itemMeta.lore[2]),
+        ),
+      );
+    }
+  },
+);
+
+async function handleMessage(msg: ChatMessage) {
+  playersWriting.delete(msg.sender);
+  const inventory = msg.sender.inventory as PlayerInventory;
+  const mainHand = inventory.itemInMainHand;
+  const offHand = inventory.itemInOffHand;
+  if (
+    offHand.type !== Material.FEATHER ||
+    mainHand.type !== Material.PAPER ||
+    mainHand?.itemMeta?.lore
+  ) {
+    msg.sender.sendActionBar('§7Kirjoittaminen peruttu.');
+    return;
+  }
+  if (!msg.sender.inventory.contains(Material.INK_SAC, mainHand.amount)) {
+    msg.sender.sendActionBar('§7Sinulla ei ole mustetta tarpeeksi.');
+    return;
+  }
+  const message = msg.content;
+  if (MAX_LENGTH < message.length) {
+    msg.sender.sendMessage(
+      color(ChatColor.GRAY, text('Tämä teksti on liian pitkä. Käytä kirjaa!')),
+    );
+    return;
+  }
+  const letter = PaperWritten.create();
+  letter.amount = mainHand.amount;
+  const itemMeta = letter.itemMeta;
+  // Add color to lore
+  itemMeta.lore = [`§7${message}`];
+  letter.itemMeta = itemMeta;
+  inventory.itemInMainHand = letter;
+  msg.sender.sendActionBar('§7Kirjoittaminen onnistui!');
+  msg.sender.playSound(
+    msg.sender.location,
+    'entity.villager.work_cartographer',
+    SoundCategory.PLAYERS,
+    1,
+    1,
+  );
+  msg.sender.inventory.removeItem(
+    new ItemStack(Material.INK_SAC, mainHand.amount),
+  );
+}
+
+/**
+ * Alternative interface for detecting writing paper hands.
+ */
+
+export function detectWritingPaper(msg: ChatMessage) {
+  if (playersWriting.has(msg.sender)) {
+    msg.discard = true;
+    handleMessage(msg);
+  }
+}
+
+GLOBAL_PIPELINE.addHandler('detectWritingPaper', 0, detectWritingPaper);
```

Comment:
Consider a negative priority to avoid triggering other handlers added with priority 0. After all, this is not really a chat message.